keps/sig-auth/2579-pod-isolation-policy/README.md

```diff
@@ -0,0 +1,951 @@
+
+[templated pods]: #podtemplate-resources
+
+<<[UNRESOLVED]>>
+
+The mode key names are open to discussion. In particular, I'm not satisfied that `allow` and
+`audit/warn` carry slightly different meanings:
+
+- `allow` means "allow pods meeting this profile level & below" (i.e. deny above)
+- `audit` and `warn` mean "audit/warn on pods exceeding this profile level" (i.e. audit/warn above)
+
+Using `deny` for the enforcing level would be more consistent with audit/warn (deny above), but is a
+counter-intuitive user experience. One would expect `deny = privileged` to deny privileged pods, but
+it really means allow privileged pods (hence using `allow` for the key).
+
+Another option is to change the `audit` and `warn` keys to something meaning "don't audit/warn at this
+level and bellow", but we haven't thought of a good name. The best alternative we've come up with is
+`audit-above` and `warn-above`, but that is cumbersome.
+
+A final option is to change `allow` to `enforce`, which is more ambiguous. The ambiguity makes the
+inconsistency with audit & warn less obvious, but doesn't help the user experience.
+
+<<[/UNRESOLVED]>>
+
+### Validation
+
+The following restrictions are placed (by the admission plugin) on the policy namespace labels:
+
+1. Unknown labels with the pod security prefix (TBD) are rejected, e.g. `podisolationpolicy.kubernetes.io/foo-bar`
+2. Policy level must be one of: `privileged`, `baseline`, `restricted`
+3. Version values must be match `(latest|v[0-9]+\.[0-9]+`. That is, one of:
+    1. `latest`
+    2. `vMAJOR.MINOR` (e.g. `v1.21`)
+
+### Versioning
+
+A specific version can be supplied for each enforcement mode. The version pins the policy to the
+version that was defined at that kubernetes version. The default version is `latest`, which can be
+provided to explicitly use the latest definition of the policy. There is some nuance to how versions
+other than latest are applied:
+
+- If the constrained pod field has not changed since the pinned version, the policy is applied as
+  originally specified.
+- The privileged profile always means fully unconstrained and is effectively unversioned (specifying
+  a version is allowed but ignored).
+- Specifying a version more recent than the current Kubernetes version is allowed (for rollback &
+  version skew reasons), but is treated as `latest`.
+
+<<[UNRESOLVED]>>
+
+Under the webhook implementation, if the webhook version is newer than the cluster version, how
+should policy versions past the cluster version be handled? Should `latest` mean the latest version
+supported by the webhook, or the policy version matching the cluster version? _Leaning towards
+supporting all policy versions supported by the webhook, even if they are newer than the cluster
+version._
+
+Note that policies are not guaranteed to be backwards compatible, and a newer restricted policy
+could require setting a field that doesn't exist in the current API version.
```

Comment:
This seems like an option for a perma-break.  However, I don't see a practical way of knowing the version of a cluster.  I don't believe we force behavior for /version for conformance purposes.

Reply:
I thought that we did resolve this? The conclusion I drew from our discussion was that under the webhook implementation, policy versions would be tied to the webhook version, not the cluster version.

Reply:
That's certainly the easiest, implementation-wise, but makes for tricky ordering on upgrade... if you upgrade the webhook first, then its `latest` restricted policy can start requiring fields to be set that the calling server might not be capable of setting yet.

I think the webhook implementation is likely secondary, so I'm ok saying that the webhook library version determines the meaning of `latest`, but we need to clearly document the expected order on upgrades between API server and webhook.